### PR TITLE
EG-1164

### PR DESCRIPTION
**Pull request by LesWilson**
Standard Date formatting for labels in format "2 January 2020"

### CODE REVIEW

File: app/models/ocelot/package.scala [hunk 1]
```diff
@@ -16,19 +16,26 @@
...
-  val inputCurrencyPoundsRegex = "^-?£?(\\d{1,3}(,\\d{3})*|\\d+)$".r
-  val integerRegex = "^\\d+$".r
-  val anyIntegerRegex = "^[\\-]?\\d+$".r
+  val pageLinkRegex: Regex = s"\\[(button|link)(-same|-tab)?:([^\\]]+?):(\\d+|${Process.StartStanzaId})\\]".r
+  val labelRefRegex: Regex = s"\\[label:([A-Za-z0-9\\s\\-_]+)(:(currency))?\\]".r
+  val inputCurrencyRegex: Regex = "^-?£?(\\d{1,3}(,\\d{3})*|\\d+)(\\.(\\d{1,2})?)?$".r
+  val inputCurrencyPoundsRegex: Regex = "^-?£?(\\d{1,3}(,\\d{3})*|\\d+)$".r
+  val integerRegex: Regex = "^\\d+$".r
+  val anyIntegerRegex: Regex = "^[\\-]?\\d+$".r
+  val dateRegex: Regex = """(29/0?2/(2000|2400|2800|(19|2\d(0[48]|[2468][048]|[13579][26]))))|
```
Comment (cmrhmrc) [inline app/models/ocelot/package.scala:35]: as discussed, we can use the LocalDate/DateFormatter implementation in EG-1160

File: app/models/ocelot/package.scala [hunk 2]
```diff
@@ -42,6 +49,11 @@ package object ocelot {
   def asInt(value: String): Option[Int] = integerRegex.findFirstIn(value).map(_.toInt)
   def asAnyInt(value: String): Option[Int] = anyIntegerRegex.findFirstIn(value).map(_.toInt)
 
+  def asDate(value: String): Option[LocalDate] =
```
Comment (cmrhmrc) [inline app/models/ocelot/package.scala:52]: as above

File: app/models/ui/OutputFormat.scala [hunk 3]
```diff
@@ -60,12 +62,26 @@ case object Txt extends OutputFormat with Name {
   val name: String = "text"
 }
 
+case object DateStandard extends OutputFormat with Name {
```
Comment (cmrhmrc) [inline app/models/ui/OutputFormat.scala:65]: we could make the output format "d MMMM uuuu" 
e.g. [label:DueDate:d MMMM uuuu] 
or    [label:DueDate:date(d MMMM uuuu)]
which would fit in the Regex,  be supported by DateFormatter and give guidance writers more flexibility

File: app/services/TextBuilder.scala [hunk 4]
```diff
@@ -29,7 +29,7 @@ object TextBuilder {
   private val answerHintPattern: Regex = """\[hint:([^\]]+)\]""".r
 
   private object Placeholders { // All the placeholder matching in one place
-    val labelPattern = "\\[label:([A-Za-z0-9\\s\\-_]+)(:(currency|currencyPoundsOnly))?\\]"
+    val labelPattern = "\\[label:([A-Za-z0-9\\s\\-_]+)(:(currency|currencyPoundsOnly|date))?\\]"
```
Comment (cmrhmrc) [inline app/services/TextBuilder.scala:32]: see above

File: app/models/ui/OutputFormat.scala (hunk 3 above)
Comment (LesWilson) [reply app/models/ui/OutputFormat.scala:65]: let's discuss as not sure we want designers to have to know the formats we use

File: test/views/components/RenderFormSpec.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import views.html.components
+
+import scala.collection.JavaConverters._
+
+class RenderFormSpec extends ViewSpec with ViewFns with GuiceOneAppPerSuite {
+  private trait Test {
+    implicit val labels: Labels = LabelCache()
+    private def injector: Injector = app.injector
+    def messagesApi: MessagesApi = injector.instanceOf[MessagesApi]
+//    def formProvider: SubmittedAnswerFormProvider = injector.instanceOf[SubmittedAnswerFormProvider]
```
Comment (woollett-r) [inline test/views/components/RenderFormSpec.scala:41]: Remove commented out code

File: test/views/components/RenderFormSpec.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      h1.text() shouldBe "Bank balance?"
+      val input: Element = getSingleElementByTag(html, "Input")
+      input.id() shouldBe "currency-0"
+    }
+
+    "render three text inputs for a DateInput" in new Test {
+      val dateInput: DateInput = models.ui.DateInput(Text("Date of birth?", "Welsh, Date of birth?"), None, Seq.empty)
+      val page: FormPage = models.ui.FormPage("/url", dateInput)
+      implicit val ctx: PageContext = models.PageContext(page, None, "sessionId", None, Text(), "processId", "processCode", labels)
+      val html: Html = components.render_form(dateInput, "dateOfBirth", formProvider("test" -> nonEmptyText))
```
Comment (woollett-r) [inline test/views/components/RenderFormSpec.scala:75]: In the case of date input the form provider is actually SubmittedDateAnswerFormProvider. So I would use

val formProvider = new SubmittedDateAnswerFormProvider()
val form = formProvider()

In the code this is done programmatically, but in the tests you have to create the form provider manually.

File: test/views/components/RenderFormSpec.scala (hunk 5 above)
Comment (LesWilson) [reply test/views/components/RenderFormSpec.scala:41]: will remove

File: test/views/components/RenderFormSpec.scala (hunk 6 above)
Comment (LesWilson) [reply test/views/components/RenderFormSpec.scala]: updated

**Pull request merged**